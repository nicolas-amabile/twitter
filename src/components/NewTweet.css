.new-tweet {
  border: solid 0.0625rem #eaeaea;
  padding: 0.75rem;
  margin: 0;
}

.new-tweet__input-wrapper {
  display: grid;
  grid-template-columns: 50px 1fr;
}

.new-tweet-input {
  flex: 1;
  padding: 0 10px;
  border: none;
  font-weight: 600;
  font-size: 1rem;
}

.new-tweet-input:focus {
  outline: none;
}

.new-tweet-input::placeholder {
  color: #9097a4;
}

.new-tweet__footer {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding-left: 60px;
}

.new-tweet__footer img {
  height: 1.25rem;
}

.new-tweet-button {
  border-radius: 20px;
  background-color: #79d1fc;
  color: white;
  height: 2.25rem;
  padding: 0 0.75rem;
  cursor: pointer;
  border: none;
  font-weight: 600;

  transition: all 0.3s ease-in-out;
}

.new-tweet-button:disabled {
  background-color: #e3e3e3;
}
