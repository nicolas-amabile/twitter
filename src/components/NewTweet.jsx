import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addTweet as addTweetAction } from '../actions';
import Avatar from './Avatar';
import { isEmpty } from '../utils';

const MAX_CHARS = 60;

export class NewTweet extends Component {

  constructor(props) {
    super(props);
    this.state = {
      text: '',
      addTweetButtonDisabled: true,
    };
    this.onTweetInputChange = this.onTweetInputChange.bind(this);
    this.getMaxCharsAllowed = this.getMaxCharsAllowed.bind(this);
  }

  getMaxCharsAllowed() {
    return this.props.maxChars ? this.props.maxChars : MAX_CHARS;
  }

  publishTweet() {
    const { user, addTweet } = this.props;
    const { text } = this.state;
    addTweet({
      content: text,
      userId: user.id,
      likes: [],
      date: new Date(),
      retweets: 0,
    });
  }

  onTweetInputChange(event) {
    
    const { value: text } = event.target;

    const addTweetButtonDisabled = text.length === 0 || text.length > this.getMaxCharsAllowed();

    this.setState({ text, addTweetButtonDisabled });
  }

  render() {
    const { user } = this.props;
    if (!user || isEmpty(user)) {
      return null;
    }
    return (
      <div className="new-tweet">
        <Avatar src={user.avatar} />
        <input
          className="new-tweet-input"
          placeholder="What's happening?"
          data-testid="new-tweet-input"
          onChange={this.onTweetInputChange}
        />
        <button
          className="new-tweet-button"
          type="button"
          data-testid="new-tweet-button"
          onClick={this.publishTweet}
          disabled={this.state.addTweetButtonDisabled}
          title={this.state.addTweetButtonDisabled ? `You need at least one character to proceed` : ''}
        >
          Tweet
        </button>
      </div>
    );
  }
}

NewTweet.propTypes = {
  user: PropTypes.object.isRequired,
  addTweet: PropTypes.func.isRequired,
};

const mapStateToProps = ({ user }) => ({
  user,
});

export default connect(mapStateToProps, { addTweet: addTweetAction })(NewTweet);
