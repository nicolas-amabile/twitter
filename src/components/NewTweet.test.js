import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { NewTweet } from './NewTweet';

const user = {
  id: 1,
  name: 'Jeremy',
  avatar: { src: '/avatar' }
}

const addTweet = () => {
  console.log('Adding Tweet')
}

describe('New tweet', () => {
  test('it renders correctly', () => {
    const { container } = render(<NewTweet user={user} addTweet={addTweet} />);
    expect(container).toMatchSnapshot();
  });
  
  test('Enforce max length for tweet by using the maxChars prop to 30 chars', () => {
    const { getByTestId } = render(<NewTweet maxChars={30} user={user} addTweet={addTweet} />);
    
    const input = getByTestId('new-tweet-input');
    const button = getByTestId('new-tweet-button');

    fireEvent.change(input, { target: {value: 'This is my first tweet with more than 30 chars'}});

    expect(button).toHaveProperty('disabled', true);

    fireEvent.change(input, { target: {value: 'Less than 30 chars'}});

    expect(button).toHaveProperty('disabled', false);

  });

  test("Enforce max length for tweet, using default value", () => {

    const { getByTestId } = render(<NewTweet user={user} addTweet={addTweet} />);
    
    const input = getByTestId('new-tweet-input');

    const button = getByTestId('new-tweet-button');

    fireEvent.change(input, { target: {value: 'This is my first tweet with more than 60 chars. This is my first tweet with more than 60 chars.'}});

    expect(button).toHaveProperty('disabled', true);

    fireEvent.change(input, { target: {value: 'Less than 60 chars'}});

    expect(button).toHaveProperty('disabled', false);

  });

});
