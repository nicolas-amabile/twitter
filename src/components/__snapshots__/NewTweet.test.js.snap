// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New tweet it renders correctly 1`] = `
<div>
  <div
    class="new-tweet"
  >
    <img
      alt="avatar"
      class="avatar"
      src="[object Object]"
    />
    <input
      class="new-tweet-input"
      data-testid="new-tweet-input"
      placeholder="What's happening?"
    />
    <button
      class="new-tweet-button"
      data-testid="new-tweet-button"
      disabled=""
      title="You need at least one character to proceed"
      type="button"
    >
      Tweet
    </button>
  </div>
</div>
`;
